### PR TITLE
Histogram rollout: <LoadTOFRawNexus to SaveDaveGrp>

### LINKED ISSUE #18621: HistogramData rollout: algorithms <LoadTOFRawNexus to SaveDaveGrp>
This issue is part of the HistogramData rollout, see https://github.com/mantidproject/mantid/issues/17641.
Algorithms covered in this issue are:
Framework/DataHandling/src/LoadTOFRawNexus.cpp
Framework/DataHandling/src/LoadVulcanCalFile.cpp
Framework/DataHandling/src/MaskDetectors.cpp
~~Framework/DataHandling/src/PDLoadCharacterizations.cpp~~
Framework/DataHandling/src/SaveAscii2.cpp
Framework/DataHandling/src/SaveAscii.cpp
Framework/DataHandling/src/SaveCanSAS1D2.cpp
Framework/DataHandling/src/SaveCanSAS1D.cpp
Framework/DataHandling/src/SaveCSV.cpp
Framework/DataHandling/src/SaveDaveGrp.cpp

### PR DESCRIPTION
**Pull request by LamarMoore**
HistogramData rollout: algorithms <X to Y>

## Description of work.
This PR is part of refactoring effort to use the new interface of the HistogramData module.
See https://github.com/mantidproject/mantid/issues/17641 for details.

The following algorithms were refactored to use HistogramData:

Framework/DataHandling/src/LoadTOFRawNexus.cpp
Framework/DataHandling/src/LoadVulcanCalFile.cpp
Framework/DataHandling/src/MaskDetectors.cpp
Framework/DataHandling/src/SaveAscii2.cpp
Framework/DataHandling/src/SaveAscii.cpp
Framework/DataHandling/src/SaveCanSAS1D2.cpp
Framework/DataHandling/src/SaveCanSAS1D.cpp
Framework/DataHandling/src/SaveCSV.cpp
Framework/DataHandling/src/SaveDaveGrp.cpp

## To test.

Code review.

Fixes #18621.
Internal change, no release notes (potential performance improvements will be added to release notes as part of the umbrella issue).

### CODE REVIEW

File: Framework/DataHandling/src/SaveAscii2.cpp [hunk 1]
```diff
@@ -301,24 +301,22 @@ void SaveAscii2::writeSpectra(const std::set<int>::const_iterator &spectraItr,
   for (int bin = 0; bin < m_nBins; bin++) {
     if (!m_isCommonBins) // checking for ragged workspace
     {
-      file << (m_ws->readX(*spectraItr)[bin] +
-               m_ws->readX(*spectraItr)[bin + 1]) /
-                  2;
+      file << (m_ws->x(*spectraItr)[bin] + m_ws->x(*spectraItr)[bin + 1]) / 2;
```
Comment (SimonHeybrock) [inline Framework/DataHandling/src/SaveAscii2.cpp]: Use `points`?

File: Framework/DataHandling/src/SaveAscii2.cpp [hunk 2]
```diff
@@ -301,24 +301,22 @@ void SaveAscii2::writeSpectra(const std::set<int>::const_iterator &spectraItr,
...
     }
 
     else if (m_isHistogram & m_isCommonBins) // bin centres,
     {
-      file << (m_ws->readX(0)[bin] + m_ws->readX(0)[bin + 1]) / 2;
+      file << (m_ws->x(0)[bin] + m_ws->x(0)[bin + 1]) / 2;
```
Comment (SimonHeybrock) [inline Framework/DataHandling/src/SaveAscii2.cpp]: `points`?

File: Framework/DataHandling/src/SaveAscii2.cpp [hunk 3]
```diff
@@ -347,21 +345,19 @@ void SaveAscii2::writeSpectra(const int &spectraIndex, std::ofstream &file) {
   for (int bin = 0; bin < m_nBins; bin++) {
     if (m_isHistogram & m_isCommonBins) // bin centres,
     {
-      file << (m_ws->readX(0)[bin] + m_ws->readX(0)[bin + 1]) / 2;
+      file << (m_ws->x(0)[bin] + m_ws->x(0)[bin + 1]) / 2;
```
Comment (SimonHeybrock) [inline Framework/DataHandling/src/SaveAscii2.cpp]: `points`?

File: Framework/DataHandling/test/SaveCSVTest.h [hunk 4]
```diff
@@ -130,11 +129,12 @@ class SaveCSVTest : public CxxTest::TestSuite {
...
     for (size_t j = 0; j < nSpec; ++j) {
+      auto &X = ws->mutableX(j);
       for (size_t k = 0; k < nBins + 1; ++k) {
-        ws->dataX(j)[k] = double(k);
+        X[k] = double(k);
       }
-      ws->dataY(j).assign(nBins, double(j));
-      ws->dataE(j).assign(nBins, sqrt(double(j)));
+      ws->mutableY(j).assign(nBins, double(j));
+      ws->mutableE(j).assign(nBins, sqrt(double(j)));
```
Comment (SimonHeybrock) [inline Framework/DataHandling/test/SaveCSVTest.h]: Use `LinearGenerator` for X and `setHistogram` to get E automatically?

**Review by SimonHeybrock — COMMENTED** [review]
Only minor comments:

File: Framework/DataHandling/src/SaveAscii2.cpp [hunk 5]
```diff
@@ -298,15 +295,17 @@ void SaveAscii2::writeSpectra(const std::set<int>::const_iterator &spectraItr,
...
+  auto points0 = m_ws->points(0);
+  auto pointsSpec = m_ws->points(*spectraItr);
   for (int bin = 0; bin < m_nBins; bin++) {
     if (!m_isCommonBins) // checking for ragged workspace
     {
-      file << (m_ws->x(*spectraItr)[bin] + m_ws->x(*spectraItr)[bin + 1]) / 2;
+      file << pointsSpec[bin];
     }
 
     else if (m_isHistogram & m_isCommonBins) // bin centres,
```
Comment (SimonHeybrock) [inline Framework/DataHandling/src/SaveAscii2.cpp]: Do you still need `m_isHistogram` (and the corresponding `if`/`else`?)?

File: Framework/DataHandling/src/SaveAscii2.cpp [hunk 6]
```diff
@@ -342,13 +341,15 @@ void SaveAscii2::writeSpectra(const int &spectraIndex, std::ofstream &file) {
   file << '\n';
 
   auto pointDeltas = m_ws->pointStandardDeviations(0);
+  auto points0 = m_ws->points(0);
+  auto pointsSpec = m_ws->points(spectraIndex);
   for (int bin = 0; bin < m_nBins; bin++) {
     if (m_isHistogram & m_isCommonBins) // bin centres,
```
Comment (SimonHeybrock) [inline Framework/DataHandling/src/SaveAscii2.cpp]: `m_isHistogram` not needed?

File: Framework/DataHandling/src/SaveAscii2.cpp [hunk 7]
```diff
@@ -306,14 +302,10 @@ void SaveAscii2::writeSpectra(const std::set<int>::const_iterator &spectraItr,
       file << pointsSpec[bin];
     }
 
-    else if (m_isHistogram & m_isCommonBins) // bin centres,
+    else if (m_isCommonBins) // bin centres,
```
Comment (SimonHeybrock) [inline Framework/DataHandling/src/SaveAscii2.cpp]: Just `else`, and comment should go.

File: Framework/DataHandling/src/SaveAscii2.cpp [hunk 8]
```diff
@@ -347,15 +339,12 @@ void SaveAscii2::writeSpectra(const int &spectraIndex, std::ofstream &file) {
   auto points0 = m_ws->points(0);
   auto pointsSpec = m_ws->points(spectraIndex);
   for (int bin = 0; bin < m_nBins; bin++) {
-    if (m_isHistogram & m_isCommonBins) // bin centres,
+    if (m_isCommonBins) // bin centres,
     {
       file << points0[bin];
     } else if (!m_isCommonBins) // checking for ragged workspace
```
Comment (SimonHeybrock) [inline Framework/DataHandling/src/SaveAscii2.cpp]: `else`

**Comment by LamarMoore** [comment]
For some reason my clang format tool in MSVC doesn't agree with the tool here.

File: Framework/DataHandling/src/SaveAscii.cpp [hunk 9]
```diff
@@ -204,25 +204,25 @@ void SaveAscii::exec() {
   for (int bin = 0; bin < nBins; bin++) {
     if (isHistogram) // bin centres
```
Comment (SimonHeybrock) [inline Framework/DataHandling/src/SaveAscii.cpp]: Use `points`?

**Comment by SimonHeybrock** [comment]
Code looks good now, :shipit:

**Pull request merged**